app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala

```diff
@@ -0,0 +1,119 @@
+import uk.gov.hmrc.plasticpackagingtax.registration.views.html.confirm_address
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ContactDetailsConfirmAddressController @Inject() (
+  authenticate: AuthAction,
+  journeyAction: JourneyAction,
+  override val registrationConnector: RegistrationConnector,
+  incorpIdConnector: IncorpIdConnector,
+  mcc: MessagesControllerComponents,
+  page: confirm_address
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with Cacheable with I18nSupport {
+
+  def displayPage(): Action[AnyContent] =
+    (authenticate andThen journeyAction).async { implicit request =>
+      request.registration.incorpJourneyId match {
+        case Some(journeyId) =>
+          for {
+            incorporationDetails <- incorpIdConnector.getDetails(journeyId)
+          } yield Ok(
+            page(form().fill(
+                   ConfirmAddress(request.registration.primaryContactDetails.useRegisteredAddress)
+                 ),
+                 incorporationDetails.companyAddress
+            )
+          )
+        case _ => Future(Redirect(routes.RegistrationController.displayPage()))
+      }
+    }
+
+  def submit(): Action[AnyContent] =
+    (authenticate andThen journeyAction).async { implicit request =>
+      request.registration.incorpJourneyId match {
+        case Some(journeyId) =>
+          incorpIdConnector.getDetails(journeyId).flatMap { incorporationDetails =>
+            form()
+              .bindFromRequest()
+              .fold(
+                (formWithErrors: Form[ConfirmAddress]) =>
+                  Future(BadRequest(page(formWithErrors, incorporationDetails.companyAddress))),
+                confirmAddress =>
+                  updateRegistration(confirmAddress, incorporationDetails.companyAddress).map {
+                    case Right(_) =>
+                      FormAction.bindFromRequest match {
+                        case SaveAndContinue =>
+                          if (confirmAddress.useRegisteredAddress.getOrElse(false))
+                            Redirect(routes.ContactDetailsCheckAnswersController.displayPage())
+                          else Redirect(routes.ContactDetailsAddressController.displayPage())
+                        case _ => Redirect(routes.RegistrationController.displayPage())
+                      }
+                    case Left(error) => throw error
+                  }
+              )
+          }
+        case _ => Future(Redirect(routes.RegistrationController.displayPage()))
+      }
+    }
+
+  private def updateRegistration(
+    formData: ConfirmAddress,
+    incorporationAddressDetails: IncorporationAddressDetails
+  )(implicit req: JourneyRequest[AnyContent]): Future[Either[ServiceError, Registration]] =
+    update { registration =>
+      val updatedPrimaryContactDetails = {
+        if (formData.useRegisteredAddress.getOrElse(false))
```

Comment:
same here, maybe we can use Option's map func?

Reply:
No because we can only redirect to `routes.ContactDetailsCheckAnswersController.displayPage()` if `true`